Controller Forced Reconfiguration API

joe-redpanda opened the pull request:
This PR adds the infrastructure for the controller forced reconfiguration api.
Namely, this PR:
1. adds CFR to the adminv2/internal/debug api
2. adds a feature barrier for it
3. wires the debug api into the controller in preparation for feature implementation

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Features

* Adds controller force reconfigure api
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda committed:
adminv2: make internal breakglass service add CFR

Adds a new internal service for breakglass apis to adminv2, adds CFR to
that service.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3d8b4975-b0c2-4900-a05c-7b4b5ba99a37-->
<!--TEST RESULTS-->
<!--18e9bb03-40c4-4e52-b8cc-5e937fa661d1-->
<!--TEST RESULTS-->
<!--0038e2c2-48b4-4c81-891d-df52ad37c663-->
<!--TEST RESULTS-->
<!--16b6d07e-20ed-4530-a933-c1cd9d92ef3d-->
<!--TEST RESULTS-->
<!--60f24ed4-ddb8-4e8d-a7b3-9a4949ea1449-->
<!--TEST RESULTS-->
<!--6d1b75cd-a541-43cf-90c1-a3d17e6db274-->
<!--TEST RESULTS-->
<!--45de403f-4fd4-43e5-a63d-2cb0a2fa23e6-->
<!--TEST RESULTS-->
<!--fc3f296d-3096-421b-b871-995cfab15c6a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74024">build#74024</a></summary>


| test_class         | test_method       | test_arguments           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                       |
|:-------------------|:------------------|:-------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------|
| SimpleEndToEndTest | test_relaxed_acks | {"write_caching": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74024#0199d071-0f37-48f2-a91e-3795c17486dd | FLAKY         | 20/21    | upstream reliability is '99.52830188679245'. current run reliability is '95.23809523809523'. drift is 4.29021 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SimpleEndToEndTest&test_method=test_relaxed_acks |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74184">build#74184</a></summary>


| test_class                 | test_method   | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                           |
|:---------------------------|:--------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------|
| SchemaRegistryAutoAuthTest | test_restarts | {"move_controller_leader": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74184#0199e3fd-23c0-4b49-b89e-39f92ae9fde0 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaRegistryAutoAuthTest&test_method=test_restarts |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74297">build#74297</a></summary>


| test_class                                     | test_method                                    | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:-----------------------------------------------|:-----------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/cluster/tests/shard_placement_table_test | src/v/cluster/tests/shard_placement_table_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74297#0199ea3c-0c72-4a63-8478-855918c255bf | FAIL          | 0/1      |          |                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74349">build#74349</a></summary>


| test_class                       | test_method                                             | test_arguments                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                   |
|:---------------------------------|:--------------------------------------------------------|:-----------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndShadowIndexingTest       | test_reset_from_cloud                                   | {"cloud_storage_type": 2}                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4bf-4905-b19a-1648bcf82571 | FAIL          | 0/1      |                                                                                                                                                                          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_from_cloud                                 |
| EndToEndShadowIndexingTest       | test_reset_from_cloud                                   | {"cloud_storage_type": 1}                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4c0-4eed-a3a5-cd6c389a07f8 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_from_cloud                                 |
| NodesDecommissioningTest         | test_decommissioning_cancel_ongoing_movements           | {"cloud_topic": true}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef76-a8b9-4580-be92-3c2b7a315d25 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_cancel_ongoing_movements           |
| NodesDecommissioningTest         | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": false}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4c0-4500-8f05-64249de66e15 | FLAKY         | 11/21    | upstream reliability is '100.0'. current run reliability is '52.38095238095239'. drift is 47.61905 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest         | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef76-a8c1-4181-9050-f9a677df4cfc | FLAKY         | 19/21    | upstream reliability is '98.01980198019803'. current run reliability is '90.47619047619048'. drift is 7.54361 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest         | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                            | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4c2-40f0-9052-56e626bb0470 | FLAKY         | 10/21    | upstream reliability is '93.22033898305084'. current run reliability is '47.61904761904761'. drift is 45.60129 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest         | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                            | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef76-a8c2-4a6e-b35f-d7a28aa25f28 | FLAKY         | 15/21    | upstream reliability is '93.22033898305084'. current run reliability is '71.42857142857143'. drift is 21.79177 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest         | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": false, "tick_interval": 3600000} | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4be-407f-b70e-39fc144e713c | FLAKY         | 14/21    | upstream reliability is '91.25'. current run reliability is '66.66666666666666'. drift is 24.58333 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest         | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": false, "tick_interval": 5000}    | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4c0-4eed-a3a5-cd6c389a07f8 | FLAKY         | 9/21     | upstream reliability is '88.75'. current run reliability is '42.857142857142854'. drift is 45.89286 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest         | test_multiple_decommissions                             | {"cloud_topic": true}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef76-a8bf-45a2-9a5b-9a5d630473a3 | FLAKY         | 7/21     | upstream reliability is '63.51351351351351'. current run reliability is '33.33333333333333'. drift is 30.18018 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_multiple_decommissions                             |
| NodesDecommissioningTest         | test_recommissioning_one_of_decommissioned_nodes        | {"cloud_topic": true}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef70-a4c2-40f0-9052-56e626bb0470 | FLAKY         | 13/21    | upstream reliability is '85.14851485148515'. current run reliability is '61.904761904761905'. drift is 23.24375 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_one_of_decommissioned_nodes        |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss                              | null                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74349#0199ef76-a8b5-4dec-85c6-b886f825d1ff | FLAKY         | 17/21    | upstream reliability is '90.70321811680571'. current run reliability is '80.95238095238095'. drift is 9.75084 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss                      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74420">build#74420</a></summary>


| test_class                           | test_method                    | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-------------------------------------|:-------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync     | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74420#0199f30c-3bfc-4cbd-b4c6-16a27171e271 | FLAKY         | 19/21    | upstream reliability is '98.36660617059891'. current run reliability is '90.47619047619048'. drift is 7.89042 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkingReplicationTests        | test_replication_with_failures | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74420#0199f315-27bf-4519-919d-6973c8fa024a | FLAKY         | 10/21    | upstream reliability is '83.20493066255779'. current run reliability is '47.61904761904761'. drift is 35.58588 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74553">build#74553</a></summary>


| test_class                          | test_method                                  | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                               |
|:------------------------------------|:---------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeAdminEndpointTest           | test_get_coordinator_state                   | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a033e-17d6-4bb8-a572-10bc26ec8f95 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state         |
| DatalakeAdminEndpointTest           | test_get_coordinator_state                   | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a034a-0acb-49ee-8ff7-db7eb4d9b486 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state         |
| DatalakeAdminEndpointTest           | test_get_coordinator_state_pending           | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a033e-17d7-447b-bc0b-c047da021459 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state_pending |
| DatalakeAdminEndpointTest           | test_get_coordinator_state_pending           | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a034a-0acc-426b-a1b1-5665f7ddc0d5 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state_pending |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a033e-17d2-45b1-8388-a358620b0cb2 | FLAKY         | 18/21    | upstream reliability is '90.69767441860465'. current run reliability is '85.71428571428571'. drift is 4.98339 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer   |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a034a-0ac6-4c5b-99c0-ad2cdaab01dc | FLAKY         | 19/21    | upstream reliability is '90.69767441860465'. current run reliability is '90.47619047619048'. drift is 0.22148 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer   |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}                            | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a034a-0ac8-4324-ac8a-71837ba0bb7f | FLAKY         | 16/21    | upstream reliability is '82.29792919171676'. current run reliability is '76.19047619047619'. drift is 6.10745 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all           |
| WriteCachingFailureInjectionTest    | test_unavoidable_data_loss                   | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74553#019a034a-0ac3-468c-bbe3-d2a8ff73bead | FLAKY         | 19/21    | upstream reliability is '91.91176470588235'. current run reliability is '90.47619047619048'. drift is 1.43557 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74751">build#74751</a></summary>


| test_class                          | test_method                                             | test_arguments                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                   |
|:------------------------------------|:--------------------------------------------------------|:---------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeAdminEndpointTest           | test_get_coordinator_state                              | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e76-c2f1-4d33-9eb4-a6f5c9ef18a4 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state                             |
| DatalakeAdminEndpointTest           | test_get_coordinator_state                              | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9eb-46ce-b6e0-0ad21cc253cd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state                             |
| DatalakeAdminEndpointTest           | test_get_coordinator_state_pending                      | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e76-c2f3-4ad3-bb1f-967f1ac2101b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state_pending                     |
| DatalakeAdminEndpointTest           | test_get_coordinator_state_pending                      | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9ec-4b5d-b92f-9e362721792b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_get_coordinator_state_pending                     |
| DatalakeUpgradeTest                 | test_upload_through_upgrade                             | {"cloud_storage_type": 1, "migration_type": "recreate_table", "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9f0-4592-859b-b37a68f12fc9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeUpgradeTest&test_method=test_upload_through_upgrade                                  |
| DatalakeUpgradeTest                 | test_upload_through_upgrade                             | {"cloud_storage_type": 1, "migration_type": "rename_columns", "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9f1-44ab-8ac6-ba2042a32ef5 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeUpgradeTest&test_method=test_upload_through_upgrade                                  |
| NodesDecommissioningTest            | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                                            | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e76-c2f5-4e40-81d6-44ef640c9c98 | FLAKY         | 19/21    | upstream reliability is '98.40255591054313'. current run reliability is '90.47619047619048'. drift is 7.92637 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test                                | {"cloud_storage_type": 1, "mixed_versions": false}                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e76-c2f5-4e40-81d6-44ef640c9c98 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test                         |
| ScalingUpTest                       | test_fast_node_addition                                 | null                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9ec-4b5d-b92f-9e362721792b | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition                                            |
| ShadowLinkingRandomOpsTest          | test_node_operations                                    | {"failures": false}                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e76-c2f5-4e40-81d6-44ef640c9c98 | FLAKY         | 10/21    | upstream reliability is '36.79245283018868'. current run reliability is '47.61904761904761'. drift is -10.82659 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest          | test_node_operations                                    | {"failures": false}                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9ed-4701-94f5-76504c568445 | FLAKY         | 3/21     | upstream reliability is '36.79245283018868'. current run reliability is '14.285714285714285'. drift is 22.50674 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest          | test_node_operations                                    | {"failures": true}                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e76-c2f6-41fe-8923-2cd36d217203 | FLAKY         | 1/21     | upstream reliability is '32.72251308900523'. current run reliability is '4.761904761904762'. drift is 27.96061 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest          | test_node_operations                                    | {"failures": true}                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9ee-4d11-9641-f8d112252c3a | FLAKY         | 1/21     | upstream reliability is '32.72251308900523'. current run reliability is '4.761904761904762'. drift is 27.96061 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                          | {"use_transactions": false}                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/74751#019a0e7e-a9eb-46ce-b6e0-0ad21cc253cd | FLAKY         | 17/21    | upstream reliability is '81.67701863354037'. current run reliability is '80.95238095238095'. drift is 0.72464 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74810">build#74810</a></summary>


| test_class                                                             | test_method                                                            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:-----------------------------------------------------------------------|:-----------------------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| MuxConsumerFixtureAndSessions/MuxConsumerFixture                       | TestProduceThrottling/1                                                |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/1      |          |                |
| MuxConsumerFixtureAndSessions/MuxConsumerFixture                       | TestResetPartition/1                                                   |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/1      |          |                |
| TopicManifestDownloaderTest                                            | TestDownloadLabeledManifestWithHint                                    |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FLAKY         | 1/2      |          |                |
| src/v/cloud_topics/level_one/metastore/tests/replicated_metastore_test | src/v/cloud_topics/level_one/metastore/tests/replicated_metastore_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/1      |          |                |
| src/v/cluster/tests/decommissioning_test                               | src/v/cluster/tests/decommissioning_test                               |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/1      |          |                |
| src/v/cluster_link/replication/tests/mux_consumer_fixture_test         | src/v/cluster_link/replication/tests/mux_consumer_fixture_test         |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/1      |          |                |
| src/v/kafka/server/tests/alter_config_test                             | src/v/kafka/server/tests/alter_config_test                             |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/2      |          |                |
| src/v/kafka/server/tests/kafka_fetch_plan_rpbench_test                 | src/v/kafka/server/tests/kafka_fetch_plan_rpbench_test                 |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FLAKY         | 1/2      |          |                |
| src/v/storage/tests/compaction_e2e_test                                | src/v/storage/tests/compaction_e2e_test                                |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/1      |          |                |
| test_with_basic_consume_fixture/BasicConsumerFixture                   | TestBasicLeadershipTransfer/0                                          |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FAIL          | 0/2      |          |                |
| test_with_basic_consume_fixture/BasicConsumerFixture                   | TestBasicLeadershipTransfer/1                                          |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FLAKY         | 1/2      |          |                |
| test_with_basic_consume_fixture/BasicConsumerFixture                   | TestBasicLeadershipTransfer/2                                          |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FLAKY         | 1/2      |          |                |
| test_with_basic_consume_fixture/BasicConsumerFixture                   | TestBasicNodeRestart/0                                                 |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74810#019a11cd-d875-4661-a07a-2a0f04e3fe77 | FLAKY         | 1/2      |          |                |
</details>
<!--/body-->

dotnwat commented on src/v/redpanda/admin/services/internal/debug.cc:
```diff
@@ -131,4 +134,49 @@ debug_service_impl::stop_stress_fiber(
+      "calling controller forced reconfiguration with nodes: {}",
+      request.get_node_ids());
+
+    if (unlikely(!_controller->get_feature_table().local().is_active(
```
there really isn't any reason to use `likely/unlikely` in anything other than hot paths

dotnwat commented on src/v/redpanda/admin/services/internal/debug.cc:
```diff
@@ -131,4 +134,49 @@ debug_service_impl::stop_stress_fiber(
+          "Feature not active yet, upgrade in progress?");
+    }
+
+    static constexpr auto error_mapper = [](const std::error_code& err) {
```
seems like this should drop `static constexpr` is there a reason for having it?

joe-redpanda replied:
Is there a reason not to?

the intent was just a function-local static helper method

dotnwat replied:
in the grand scheme of things it doesn't matter and probably it gets inlined anyway, but the existence of the modifiers makes one wonder if they serve a purpose that isn't obvious. for example, there doesn't seem to be any constant expressions, but it's marked as constexpr, so it's unclear if i'm missing something. and marking it static changes it storage duration but it doesn't seem like there is any shared state or expensive initialization, so is there another reason?

dotnwat replied:
> the intent was just a function-local static helper method

if you drop `static constexpxr` then its still a function-local helper method, right?

joe-redpanda replied:
Thanks for the explanation.

Yeah theres no strong intention to the modifiers. I'll drop them for review-ability.

bharathv commented on proto/redpanda/core/admin/v2/internal/debug.proto:
```diff
@@ -43,6 +43,12 @@ message ThrowStructuredExceptionRequest {
 message ThrowStructuredExceptionResponse {}
 
+message ControllerForcedReconfigurationRequest {
+    repeated int32 node_ids = 1;
```
nit: documentation, also call it dead_brokers or something more descriptive?

joe-redpanda committed:
adminv2/internal/breakglass: make and wire stub

Adds stubbed cpp implementation of breakglass service. Wires the service
into application.

joe-redpanda committed:
adminv2/debug: wire CFR into controller

Wires the CFR rpc into a stubbed controller implementation of cfr.

involved:
adds controller to the breakglass service
adds a stubbed implemenetation of CFR to controller
    - actual implementation will come in a different commit
adds converters to map request -> CFR call & CFR call -> response

joe-redpanda committed:
features: add CFR feature

Adds a feature for CFR to feature table

joe-redpanda committed:
rptest/services: Type improvements

typing improvements to redpanda.py and admin.py

type imporvement fixup

joe-redpanda committed:
admin/v2.py: allow node specification

Current behavior selects a random running node against which to create a
client. This commit allows the caller to optionally specify a
clusternode to target with requests

joe-redpanda committed:
admin/v2.py: add breakglass service to v2

Allows v2 admin to create a breakglass service client

joe-redpanda committed:
rptest/tests: add CFR rpc smoke test

adds a ducktape test which performs the following:
1. create a cluster of size three
2. add a topic and produce to it
3. fail the majority of nodes
4. reboot into recovery mode
5. force reconfigure the cluster to the remaining survivor
6. add new brokers back to three
7. reboot into normal mode
8. produce to the topic
9. check that all partitions on the topic have a voter set of 3

joe-redpanda committed:
breakglass: generate python bindings

Generates the python bindings for the breakglass service.

joe-redpanda commented:
Changes since last review

1. split the cfr api into a standalone service for escape hatches called 'breakglass'
2. rebase to pick up new pathing for pbgen

rockwotj commented on src/v/redpanda/admin/services/internal/BUILD:
```diff
@@ -11,7 +11,9 @@ redpanda_cc_library(
     deps = [
         "//proto/redpanda/core/admin/v2/internal:debug_redpanda_proto",
         "//src/v/base",
+        "//src/v/cluster",
         "//src/v/finjector",
+        "//src/v/model",
```
Does this belong in 740a622315089da4abf9a4e6da63c676aba164db?

rockwotj commented on proto/redpanda/core/admin/v2/internal/breakglass.proto:
```diff
@@ -0,0 +1,41 @@
+// extraordinary circumstances
+//
+service BreakglassService {
+    rpc ControllerForcedReconfiguration(ControllerForcedReconfigurationRequest)
```
Please add a comment to the RPC to what it does, because "ForcedReconfiguration" is a common term for the replication team, but not others. Here's my take:

```
// Forcibly removes these node IDs from the controller replica set.
// This can be used as an escape hatch if there is no majority remaining
// of the cluster nodes to forcible remove and recover what is left of the
// cluster controller state.
//
// This request is idempotent.
//
// Must be called on all nodes in the cluster (I think you mentioned this).
```

rockwotj commented on src/v/redpanda/admin/services/internal/breakglass.cc:
```diff
@@ -1,17 +1,58 @@
+      "finished initializing controller forced reconfiguration with errc: {}",
+      err);
+
+    error_mapper(err);
```
nit: `check_err(error)` is maybe a better name?

rockwotj commented on tests/rptest/clients/admin/v2.py:
```diff
@@ -88,11 +89,19 @@ def _make_service(self, service_clazz):
+        self, node: ClusterNode | None = None
+    ) -> shadow_link_pb2_connect.ShadowLinkServiceClient:
+        return self._make_service(
+            shadow_link_pb2_connect.ShadowLinkServiceClient, node=node
```
nit: consider using `kwargs` and just passing that through to `_make_service` for future things we add here.

rockwotj reviewed (commented):
LGTM, just some nits.

joe-redpanda commented on src/v/cluster/types.h:
```diff
@@ -3687,6 +3687,13 @@ struct get_current_cluster_epoch_response
     auto serde_fields() { return std::tie(ec, epoch); }
 };
 
+struct error_info {
```
we now have merge order, this was cherry picked from the CFR PR

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74349-->
<!--bea878cb-59c9-4f1d-9120-232737351fdd-->
<!--d954a51d-4c2c-48ef-abee-4aebe56abbc7-->
<!--title-->
## Retry command for [Build#74349](https://buildkite.com/redpanda/redpanda/builds/74349)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_from_cloud@{"cloud_storage_type":2}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_from_cloud@{"cloud_storage_type":1}
```
<!--/body-->

joe-redpanda commented:
~~Note, this pr is rebased onto https://github.com/redpanda-data/redpanda/pull/27956
to get a working round of CI~~

Fixed with merge and rebase

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74452-->
<!--167d7497-1b75-4dd9-8914-8bb96b8a71c8-->
<!--title-->
## Retry command for [Build#74452](https://buildkite.com/redpanda/redpanda/builds/74452)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkBasicTests.test_rapid_shadow_link_toggling
```
<!--/body-->

bharathv commented on proto/redpanda/core/admin/v2/internal/breakglass.proto:
```diff
@@ -0,0 +1,61 @@
+    // A given broker may only accept one CFR at a time, subsequent
+    // invocations will fail.
+    //
+    // A CFR may be cancelled and retried by rebooting all brokers.
```
nit: remove CFR acronym

bharathv commented on proto/redpanda/core/admin/v2/internal/breakglass.proto:
```diff
@@ -0,0 +1,61 @@
+option (redpanda.core.pbgen.cpp_namespace) = "proto::admin::internal";
+
+message ControllerForcedReconfigurationRequest {
+    // node ids of lost brokers to be forcibly removed
```
nit: comments end with periods for doc generation.

bharathv commented on src/v/redpanda/admin/services/internal/breakglass.cc:
```diff
@@ -1,17 +1,77 @@
+    vlog(
+      logger.info,
+      "calling controller forced reconfiguration with nodes: {}",
+      request.get_dead_node_ids());
```
nit: surviving nodes.

bharathv commented on tests/rptest/tests/controller_forced_reconfiguration_test.py:
```diff
@@ -0,0 +1,328 @@
+    @cluster(num_nodes=6)
+    def test_smoke_cfr(self):
+        """
+        1. create a cluster of size three
+        2. add a topic and produce to it
+        3. fail the majority of nodes in the cluster
+        4. reboot into recovery mode
+        5. force reconfigure the cluster to the remaining survivor
+        6. add new brokers back to three
+        7. reboot into normal mode
+        8. produce to topic
+        9. check that all partitions on topic have voter set of 3
+        """
```
Few more test suggestions (that are usually problematic).

* Can we have multiple topics (mix of rf=1/3)
* Can we have some partition movement happening when stopping nodes (to ensure we can recover _all_ partitions regradless of the state they are in). (check PartitionMovementMixin to generate random moves).
* nodes already stuck in decommission.

joe-redpanda replied:
Work in progress but currently flakey. Will add in the stabilization time before release

bharathv replied:
👍

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74553-->
<!--6553e8ec-3d89-4f96-ba3a-64ae246fd25a-->
<!--9c4a6ed8-a13d-4ea6-b3c7-626ffb5fde93-->
<!--a031c28a-dba3-4906-8064-aa1be1f9d149-->
<!--c85754df-0dc0-43b1-ae33-02a30d47877e-->
<!--title-->
## Retry command for [Build#74553](https://buildkite.com/redpanda/redpanda/builds/74553)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/admin_endpoint_test.py::DatalakeAdminEndpointTest.test_get_coordinator_state@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/admin_endpoint_test.py::DatalakeAdminEndpointTest.test_get_coordinator_state_pending@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
```
<!--/body-->

rockwotj commented on proto/redpanda/core/admin/v2/internal/breakglass.proto:
```diff
@@ -0,0 +1,61 @@
+// BreakglassService contains expert (internal) only apis to be used in
+// extraordinary circumstances
+//
+service BreakglassService {
```
last nit: service should go first in proto files, with request/response, etc after

rockwotj reviewed (commented):
Proto LGTM outside of my nit about ordering. Thanks for working on this!

bharathv commented on src/v/redpanda/admin/services/internal/breakglass.cc:
```diff
@@ -0,0 +1,79 @@
+    vlog(
+      logger.info,
+      "calling controller forced reconfiguration with nodes: {} and "
+      "surviving_node_count",
```
missing placeholder.

joe-redpanda replied:
oh wow, oops

bharathv reviewed (commented):
lgtm once the fmt placeholder is fixed.

dotnwat reviewed (commented):
i think @bharathv may have had some thoughts around the implications of controller forced reconfiguration and L0 GC in cloud topics. IIRC we had decided that it would be safe, but I'm just curious if you'd thought any more about it or if my memory is wrong and it was problematic! cc @rockwotj

rockwotj commented:
> i think @bharathv may have had some thoughts around the implications of controller forced reconfiguration and L0 GC in cloud topics. IIRC we had decided that it would be safe, but I'm just curious if you'd thought any more about it or if my memory is wrong and it was problematic! cc @rockwotj 

I kind of think it's not safe right? Like the controller committed offset may move backwards and break the assertions we have that the epoch only moves forward.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74751-->
<!--8c378ba1-65c8-41a1-bfbb-5bd82acf24ce-->
<!--8b1f895c-dde9-4d10-872e-e91fbcf6bc0a-->
<!--f08c7795-93d3-48eb-96ca-3384bf9bf0ac-->
<!--a156290b-11e8-4883-a9c1-2cabcbbf4764-->
<!--30ed8b18-986e-4883-a0a8-81c928e46641-->
<!--17dc7a92-7bc0-4b25-8635-fe111291a44e-->
<!--1ea90edf-41b8-437f-aede-e82c0e61d1f3-->
<!--title-->
## Retry command for [Build#74751](https://buildkite.com/redpanda/redpanda/builds/74751)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/admin_endpoint_test.py::DatalakeAdminEndpointTest.test_get_coordinator_state@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"migration_type":"recreate_table","query_engine":"spark"}
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"migration_type":"rename_columns","query_engine":"spark"}
tests/rptest/tests/datalake/admin_endpoint_test.py::DatalakeAdminEndpointTest.test_get_coordinator_state_pending@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
```
<!--/body-->

joe-redpanda commented:
> i think @bharathv may have had some thoughts around the implications of controller forced reconfiguration and L0 GC in cloud topics. IIRC we had decided that it would be safe, but I'm just curious if you'd thought any more about it or if my memory is wrong and it was problematic! cc @rockwotj

I don't think anything is safe with controller forced reconfiguration. any partition can suffer complete replica loss, and the controller can suffer data loss

bharathv commented:
> i think @bharathv may have had some thoughts around the implications of controller forced reconfiguration and L0 GC in cloud topics. IIRC we had decided that it would be safe, but I'm just curious if you'd thought any more about it or if my memory is wrong and it was problematic! cc @rockwotj

I don't recall the exact discussion/conclusion but I think it was about computing a safe GC epoch when replicas are lagging or something along those lines.

> I kind of think it's not safe right? Like the controller committed offset may move backwards and break the assertions we have that the epoch only moves forward.

I agree that this is unsafe and likely not unique to this particular PR. It’s also possible with the KV store surgeries we perform today (there’s always a slight risk that the commit index could move backward)

@rockwotj, could you remind me which specific assertions would fail in this case?

My current mental model is that if this happens, some partitions will have observed a higher epoch. They’ll notice that epoch_service is returning a stale epoch and will request it to invalidate via invalidate_epoch_cache(int epoch_causing_monotonicity_violation). At that point, everything just freezes until a higher epoch becomes available. Is that wrong?

rockwotj commented:
> @rockwotj, could you remind me which specific assertions would fail in this case?

https://github.com/redpanda-data/redpanda/blob/d37a642a26a4ad943eb61d8030eeb87aa698ce3e/src/v/cluster/cluster_epoch_service.cc#L444-L449

> My current mental model is that if this happens, some partitions will have observed a higher epoch. They’ll notice that epoch_service is returning a stale epoch and will request it to invalidate via invalidate_epoch_cache(int epoch_causing_monotonicity_violation). At that point, everything just freezes until a higher epoch becomes available. Is that wrong?

We don't currently use the `invalidate_epoch_cache` method (we should do this), but yes the code within the ctp stm should be fine because of the fencing mechanism.

bharathv commented:
> > @rockwotj, could you remind me which specific assertions would fail in this case?
> 
> https://github.com/redpanda-data/redpanda/blob/d37a642a26a4ad943eb61d8030eeb87aa698ce3e/src/v/cluster/cluster_epoch_service.cc#L444-L449
> 

q: The epoch computation will still be monotonic in this code path right?...  After the controller recovery finishes... the base (first) epoch has moved back and every subsequent epoch would still be monotonic relative to that (so the assert shouldn't fire)? did I miss something.

> > My current mental model is that if this happens, some partitions will have observed a higher epoch. They’ll notice that epoch_service is returning a stale epoch and will request it to invalidate via invalidate_epoch_cache(int epoch_causing_monotonicity_violation). At that point, everything just freezes until a higher epoch becomes available. Is that wrong?
> 
> We don't currently use the `invalidate_epoch_cache` method (we should do this), but yes the code within the ctp stm should be fine because of the fencing mechanism.

ack thanks.. I'm thinking we should have a utility that can advance the epoch to a desired value as an escape hatch for extreme cases like this before GA.

rockwotj commented:
> q: The epoch computation will still be monotonic in this code path right?... After the controller recovery finishes... the base (first) epoch has moved back and every subsequent epoch would still be monotonic relative to that (so the assert shouldn't fire)? did I miss something.

As long as all nodes are restarted first yes, but if a node is live from before majority was lost and a new leader shows up and reports a lower committed offset, then this assertion will fire. I don't know all the steps here.

> ack thanks.. I'm thinking we should have a utility that can advance the epoch to a desired value as an escape hatch for extreme cases like this before GA.

This is a great idea. We should totally. I think we can just have a "override epoch" stored in the controller log and have some `std::min(override_epoch, committed_offset)`

bharathv commented:
> As long as all nodes are restarted first yes, but if a node is live from before majority was lost and a new leader shows up and reports a lower committed offset, then this assertion will fire. I don't know all the steps here.

That’s exactly how this is designed. Controller reconfiguration only works in recovery mode, all surviving brokers must first be restarted in recovery mode. Once the reconfiguration is complete, they need to be restarted again out of recovery mode. Since no user partitions are loaded in recovery mode, nothing attempts to probe for an epoch until after the reconfiguration is complete and the cluster is fully restarted.

rockwotj commented:
> That’s exactly how this is designed. Controller reconfiguration only works in recovery mode, all surviving brokers must first be restarted in recovery mode. Once the reconfiguration is complete, they need to be restarted again out of recovery mode. Since no user partitions are loaded in recovery mode, nothing attempts to probe for an epoch until after the reconfiguration is complete and the cluster is fully restarted.

cool, okay then I think we're safe because we only assert that in memory state moves forward monotonically. Everywhere else we handle gracefully at the moment. I just hope we don't break that ever, because that seems a bit tricky to test 😆

joe-redpanda commented:
Not sure what happened to CI here. Lots of timeouts, bazel tidy failure and unrelated test failure on direct_consumer. I'll take a look later at what went wrong. For now, retry
